contributing/building-a-bento-amp-extension.md

```diff
@@ -0,0 +1,681 @@
+mention what rules the validator needs to validate. This usually
+includes
+
+-   Your element tag-name
+-   Required attributes for the element
+-   Specific values that an attribute accepts (e.g. `myattr="TYPE1|TYPE2"`)
+-   Layouts your element supports (see [Layout specs](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md) and [Layouts supported in your element](#layouts-supported-in-your-element))
+-   If there are restrictions where your element can or can't appear (e.g. disallowed_ancestory, mandatory_parent...)
+
+For more details take a look at [Contributing Component Validator
+Rules](https://github.com/ampproject/amphtml/blob/master/contributing/component-validator-rules.md).
+
+## Performance considerations
+
+### Loading external resources
+
+If your extension needs to load external resources (like an sdk) then
+you might need to add proper third party integration for it to work and
+use the proper third party iframe. Loading external resources is only
+allowed inside a 3p iframe which AMP serves on a different domain for
+security and performance reasons. Take a look at adding
+[&lt;amp-facebook&gt;](https://github.com/ampproject/amphtml/pull/1479/files)
+extension PR for examples of 3p integration.
+
+Read about [Inclusion of third party software, embeds and services into
+AMP](https://github.com/ampproject/amphtml/blob/master/3p/README.md).
+
+For contrast, take a look at `amp-instagram` which does NOT require an SDK
+to be loaded in order to embed a post, instead it provides an
+iframe-based embedding allowing `amp-instagram` extension to use a normal
+iframe with no 3p integration needed, similarly, `amp-youtube` and others.
+
+## Layouts supported in your element
+
+AMP defines different layouts that elements can choose whether or not to
+support Your element needs to announce which layouts it supports through
+overriding the `isLayoutSupported(layout)` callback and returning true
+if the element supports that layout. [Read more about AMP Layout
+System](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md)
+and [Layout
+Types](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md#layout).
+
+### What layout should your element support?
+
+After understanding each layout type, if it makes sense, support all of
+them. Otherwise choose what makes sense to your element. A popular
+support choice is to support size-defined layouts (Fixed, Fixed Height,
+Responsive and Fill) through using the utility `isLayoutSizeDefined`
+in `layout.js`.
+
+For example, `amp-inline-gallery-thumbnails` only supports fixed-height layout.
+
+```javascript
+class AmpPixel extends BaseElement {
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.FIXED_HEIGHT;
+  }
+}
+```
+
+While `amp-base-carousel` supports all size-defined layouts.
+
+```javascript
+class AmpBaseCarousel extends AMP.BaseElement {
```

Comment:
Seeing that this content is duplicated and not Bento/Preact-specific, maybe as a follow-up TODO/FR some of these pieces could be extracted into their own `layout-support.md` and linked to from both version of this doc.